src/auth/src/credentials.rs

```diff
@@ -0,0 +1,45 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use http::header::{HeaderName, HeaderValue};
+use std::future::Future;
+
+pub type Result<T> = std::result::Result<T, crate::errors::AuthError>;
+
+/// Represents an auth credential used to obtain auth tokens.
+/// Implementors of this trait provide a way to asynchronously retrieve tokens
+/// and construct auth headers.
+pub trait Credential: Send + Sync {
+    /// Asynchronously retrieves a token.
+    ///
+    /// This function returns a `Future` that resolves to a `Result` containing
+    /// either the `Token` or an `AuthError` if an error occurred during
+    /// token retrieval.
+    fn get_token(&mut self) -> impl Future<Output = Result<crate::token::Token>> + Send;
+
+    /// Asynchronously retrieves auth headers.
+    ///
+    /// This function returns a `Future` that resolves to a `Result` containing
+    /// either a vector of key-value pairs representing the headers or an
+    /// `AuthError` if an error occurred during header construction.
+    fn get_headers(
```

Comment:
nit: this seems associated with a token rather than the credential itself right?

Reply:
It is broader than just token. Right now possible headers are:
- Authorization
- x-goog-api-key
- x-goog-user-project
- x-goog-api-client
- google-byoid-sdk

Future header:
x-allowed-locations

Reply:
> is that what you were thinking about?

Yes, I think these can be consolidated.

Reply:
Yes, I had same thought. The comment I put this in got outdated. Here it is

> On second thought, I cannot think of a use case where someone want to know the quota project or token alone. The get_headers() should suffice for the client library to consume auth lib I believe.
> 
> We still need get_universe_domain() for the client lib to determine which universe to use to build the urls.
> 
> So I am removing get_token and get_quota_project for now. We can add them later if we find a need for it.
> 
> LMK if either of you think client lib will need those 2 methods. @codyoss and @coryan